File: leetcode/Minimum Window Substring.go
/*
Given two strings s and t of lengths m and n respectively, return the minimum window 
substring
 of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.

 

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
Example 2:

Input: s = "a", t = "a"
Output: "a"
Explanation: The entire string s is the minimum window.
Example 3:

Input: s = "a", t = "aa"
Output: ""
Explanation: Both 'a's from t must be included in the window.
Since the largest window of s only has one 'a', return empty string.
 

Constraints:

m == s.length
n == t.length
1 <= m, n <= 105
s and t consist of uppercase and lowercase English letters.
 

Follow up: Could you find an algorithm that runs in O(m + n) time?

*/

package main

import (
    "fmt"
)

func minWindow(s string, t string) string {
    if len(t) == 0 || len(s) == 0 {
        return ""
    }

    charsNeeded := make(map[byte]int)
    for i := 0; i < len(t); i++ {
        charsNeeded[t[i]]++
    }

    uniqueCharsToFind := len(charsNeeded)
    var left, right, formed int
    charsInWindow := make(map[byte]int)
    bestWindow := [3]int{1<<31 - 1, 0, 0}

    for right < len(s) {
        character := s[right]
        charsInWindow[character]++

        if count, exists := charsNeeded[character]; exists && charsInWindow[character] == count {
            formed++
        }

        for formed == uniqueCharsToFind {
            if right-left+1 < bestWindow[0] {
                bestWindow = [3]int{right - left + 1, left, right}
            }

            character = s[left]
            charsInWindow[character]--
            if count, exists := charsNeeded[character]; exists && charsInWindow[character] < count {
                formed--
            }
            left++
        }

        right++
    }

    if bestWindow[0] == 1<<31-1 {
        return ""
    }
    return s[bestWindow[1]:bestWindow[2]+1]
}

func main() {
    fmt.Println(minWindow("ADOBECODEBANC", "ABC"))  // Output: "BANC"
    fmt.Println(minWindow("a", "a"))                 // Output: "a"
    fmt.Println(minWindow("a", "aa"))                // Output: ""
    fmt.Println(minWindow("cabwefgewcwaefgcf", "cae")) // Output: "cwae"
    fmt.Println(minWindow("ab", "b"))                // Output: "b"
}

File: leetcode/Minimum Window Substring.py
"""
Given two strings s and t of lengths m and n respectively, return the minimum window 
substring
 of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.

 

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
Example 2:

Input: s = "a", t = "a"
Output: "a"
Explanation: The entire string s is the minimum window.
Example 3:

Input: s = "a", t = "aa"
Output: ""
Explanation: Both 'a's from t must be included in the window.
Since the largest window of s only has one 'a', return empty string.
 

Constraints:

m == s.length
n == t.length
1 <= m, n <= 105
s and t consist of uppercase and lowercase English letters.
 

Follow up: Could you find an algorithm that runs in O(m + n) time?
"""

from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not t or not s:
            return ""

        # Count how many of each character we need from t
        chars_needed = Counter(t)
        # Number of unique characters we need to find in s
        unique_chars_to_find = len(chars_needed)

        # Window boundaries
        window_start = 0
        window_end = 0
        # Count of characters in our current window
        chars_in_window = {}
        
        # Best solution so far: (length, start index, end index)
        best_window = float("inf"), None, None

        while window_end < len(s):
            # Add one character from s to our window
            character = s[window_end]
            chars_in_window[character] = chars_in_window.get(character, 0) + 1

            # If we've found all instances of this character we need
            if character in chars_needed and chars_in_window[character] == chars_needed[character]:
                unique_chars_to_find -= 1

            # Try to contract the window until it's no longer valid
            while unique_chars_to_find == 0:
                character = s[window_start]

                # Update if this window is the smallest so far
                if window_end - window_start + 1 < best_window[0]:
                    best_window = (window_end - window_start + 1, window_start, window_end)

                # Remove the character from the start of our window
                chars_in_window[character] -= 1
                if character in chars_needed and chars_in_window[character] < chars_needed[character]:
                    unique_chars_to_find += 1

                window_start += 1    

            # Expand the window by moving the right pointer
            window_end += 1    

        # If no valid window was found
        return "" if best_window[0] == float("inf") else s[best_window[1] : best_window[2] + 1]



# Test Cases
sol = Solution()

print(sol.minWindow("ADOBECODEBANC", "ABC"))  # Output: "BANC"
print(sol.minWindow("a", "a"))                # Output: "a"
print(sol.minWindow("a", "aa"))               # Output: ""
print(sol.minWindow("cabwefgewcwaefgcf", "cae"))  # Output: "cwae"
print(sol.minWindow("ab", "b"))               # Output: "b"
